### ISSUE TITLE
Selenium/Capybara spec infrastructure does not properly support javascript driver testing

### ISSUE THREAD

**Issue by thewoolleyman** [shakacode/react-webpack-rails-tutorial#189]
Hi,

Just giving you a heads up that I had to do some work to get full interactive-page Capybara javascript testing working with react.

Your Capybara specs for the current form submission submit-button example works, even without JS support, but when I made (in my app) a dynamic form that updated the page based on input onchange (keyup) events in a form field, it didn't work.  Here's a summary of the changes I had to make, with links to the relevant commits in my app.

I'm currently too busy with my app to make a pull request, but wanted to report this in case other people following this template project had similar issues.

There may have been a way to fix this otherwise, by ensuring the prod assets were served, but this is what I ended up with.

Also, I'm on Rails 5 and pretty much latest versions of all other major libs (including rspec), so that may have made a difference too.

Here's per-line links to the relevant commits in my app (it's diverged form the example quite a bit, but you should get the point):

* Here's the handlechange call (in a react-bootstrap input element) that didn't work under Capybara, setting me down the path to all the other required fixes:  https://github.com/deploysage/deploysage/commit/ecec378e292d4f421e30bdaafd35f514ac55aa13#diff-d2810ff8e090f9219b6e73bff0655c87R60
* Here's the change to application_helpers to treat the Rails test env as using dev, not prod, assets:  https://github.com/deploysage/deploysage/commit/ecec378e292d4f421e30bdaafd35f514ac55aa13#diff-127a850b7d7b20173ad3fba88f8a25e3L4
* You have to start the 'server-rails-hot' client dev server to properly serve dev assets in addition to the other two you already start - so I had to scrap the "ensure_assets_compiled" and just blow up if all three processes aren't already running (this could be automated again, but I didn't yet) https://github.com/deploysage/deploysage/commit/ecec378e292d4f421e30bdaafd35f514ac55aa13#diff-8d27918d1541bfa15c0e825c61d78941R18
* ...- I added it to Procfile.spec and wrapped in a bin/start-spec script for convenience: https://github.com/deploysage/deploysage/commit/ecec378e292d4f421e30bdaafd35f514ac55aa13#diff-ef9059a5217adfd50e9cdb2d7a1595a0R5
* You have to use the Capybara javascript driver.  You could probably do per-example `:js => true`, but I have a Single Page App, so I just made the javascript driver the default: https://github.com/deploysage/deploysage/commit/ecec378e292d4f421e30bdaafd35f514ac55aa13#diff-e11bdaf9116be3911e9557b69e242370R78
* Finally, there's a bug with Capybara's `fill_in` support with React.  Here's the workaround, with a link to the discussion on the Capybara issues page: https://github.com/deploysage/deploysage/commit/ecec378e292d4f421e30bdaafd35f514ac55aa13#diff-e11bdaf9116be3911e9557b69e242370R89

Anyway, hope this helps somebody.  Took me a while to figure it all out.

-- Chad

**Comment by justin808** [comment]
@thewoolleyman Thanks! This is super helpful. Can you summarize the differences with Rails 5? Would it make sense to update the tutorial app to Rails 5? Maybe a branch would be nice to see what's required.

**Comment by thewoolleyman** [comment]
@justin808 For this particular topic, I don't know that there's any differences that matter between Rails 4 and Rails 5.

The easiest thing would be to just update the tutorial app to do a field onchange event like I did, and try to test it.  For example, just have some "preview" area below the form that auto-updates with whatever you've currently typed in the field (pretending it's markdown but really just the same text).

Then write a capybara spec that tests it, using fill_in and then asserting on the preview text with what was filled in.

Either that doesn't work (as I found), which means the same issues exist in Rails 4, or it does work, meaning either I did something dumb when basing my app on the example, or something's different about my app/env/setup/etc (Rails 5 or something else).

If I get time I could try to hack on it, but don't count on it for a while, my hack time is devoted to moving my app forward and meeting some milestones on it.  Maybe after that...

Glad to answer questions, though, and give references to what I have working in my app.  

-- Chad

**Comment by justin808** [comment]
@thewoolleyman We do have capybara tests working with React. So I'm not sure what exactly is the issue that you're describing. Be sure to check out: https://github.com/shakacode/react_on_rails/blob/npm-react-on-rails-js/spec/dummy/spec/features/integration_spec.rb

I've set this up to allow changing between 4 JS drivers: https://github.com/shakacode/react_on_rails/blob/npm-react-on-rails-js/spec%2Fdummy%2Fspec%2Frails_helper.rb

The only "trick" was that es5-shim/sham is included in the webpack config:
https://github.com/shakacode/react_on_rails/blob/npm-react-on-rails-js/spec%2Fdummy%2Fclient%2Fwebpack.client.js#L5

**Comment by thewoolleyman** [comment]
OK, I forgot the generator had tests like this.

I see that you're using #set instead of #fill_in (https://github.com/shakacode/react_on_rails/blob/npm-react-on-rails-js/spec/dummy/spec/features/integration_spec.rb#L10)

I'm using something very similar to the component that the generator generates, but #set didn't work for me, and #fill_in didn't either without changes.

I thought that #set doesn't call event handling, so I'm curious how it works at all since the name is updated via the onchange event.

I'll have to generate an app and try to see what's different between that and my app.

**Comment by thewoolleyman** [comment]
Hmm, well apparently most of the changes I had to make are now not necessary to make my Capybara tests pass, and #fill_in and find by label works too.

This is strange, because prior to committing, I disabled my changes one by one and verified that the tests failed without them.  Either I was confused with the interactions between things, or there was something else different about the environment at that time causing the failures.  There's a lot of moving parts to this ;)

The one thing that is still confusing is the behavior of  #save_and_open_page.  If you call that AFTER filling in a field with the new value, the opened page does NOT have the field value change, even though the H3 text is updated with the new text, and the prod assets return 404's in the JS console, even if the Procfile.spec is running.  That's what originally led me down the wrong path, I think.

But, the captured failure screen shows show the correct value in the field, so you just have to be aware that save_and_open_page is basically useless.  It might even be good to instrument it to spit out a warning if someone inadvertently attempts to use it.

In any case, seems to be working now with the similar setup as the generator creates, so I'll close this issue.

Sorry for the noise...
-- Chad
